src/TimerContext.tsx

```diff
@@ -0,0 +1,184 @@
+// Import necessary React functions and types
+import { createContext, useContext, useState, useEffect } from "react";
+import type { ReactNode } from "react"; // Importing ReactNode type for children
+
+// ------------------- Timer Type Definition -------------------
+
+// Defining the structure of a Timer object
+export type Timer = {
+  id: string; // A unique identifier for the timer
+  type: "stopwatch" | "countdown" | "XY" | "tabata"; // Type of timer
+  duration: number; // Total duration of the timer in milliseconds
+  status: "not running" | "running" | "paused" | "completed"; // Current status of the timer
+};
+
+// ------------------- Timer Context Type -------------------
+
+// Describes the structure of the TimerContext
+type TimerContextType = {
+  timers: Timer[]; // Array of all timers
+  currentTimerIndex: number | null; // Index of the currently active timer
+  toggleStartPause: () => void; // Function to start/pause the active timer
+  fastForward: () => void; // Function to complete the current timer and move to the next
+  addTimer: (timer: Timer) => void; // Function to add a new timer
+  resetTimers: () => void; // Function to reset all timers
+};
+
+// ------------------- Creating Timer Context -------------------
+
+// Creating the TimerContext with an initial value of undefined
+const TimerContext = createContext<TimerContextType | undefined>(undefined);
+
+// ------------------- TimerProvider Component -------------------
+
+// TimerProvider wraps the app to provide the TimerContext
+export const TimerProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
+  // State to hold all timers
+  const [timers, setTimers] = useState<Timer[]>([]);
+
+  // State to track the index of the currently active timer
+  const [currentTimerIndex, setCurrentTimerIndex] = useState<number | null>(null);
+
+  // ------------------- Timer Logic (useEffect) -------------------
+
+  // Handles updating the active timer's state and duration
+  useEffect(() => {
+    if (currentTimerIndex === null) return; // Do nothing if no timer is active
+
+    const currentTimer = timers[currentTimerIndex]; // Get the active timer
+
+    if (currentTimer?.status === "running") {
+      // If the timer is running, start an interval
+      const interval = setInterval(() => {
```

Comment:
this will get executed every second, use `setTimeout` instead as explained in lecture

Reply:
Implemented!